File: src/main/java/cn/enaium/joe/dialog/search/SearchMethodDialog.java
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.dialog.search;

import cn.enaium.joe.JavaOctetEditor;
import cn.enaium.joe.dialog.SearchDialog;
import cn.enaium.joe.gui.panel.search.ResultNode;
import cn.enaium.joe.task.SearchFieldTask;
import cn.enaium.joe.task.SearchMethodTask;
import cn.enaium.joe.util.ASyncUtil;
import cn.enaium.joe.util.LangUtil;
import cn.enaium.joe.util.MessageUtil;
import cn.enaium.joe.util.StringUtil;
import org.objectweb.asm.tree.*;

import javax.swing.*;
import java.awt.*;

/**
 * @author Enaium
 * @since 0.5.0
 */
public class SearchMethodDialog extends SearchDialog {
    public SearchMethodDialog() {
        setTitle(LangUtil.i18n("search.method.title"));
        add(new JPanel(new FlowLayout()) {{
            add(new JLabel(LangUtil.i18n("search.owner")));
            JTextField owner = new JTextField();
            add(owner);
            add(new JLabel(LangUtil.i18n("search.name")));
            JTextField name = new JTextField();
            add(name);
            add(new JLabel(LangUtil.i18n("search.description")));
            JTextField description = new JTextField();
            add(description);
            JCheckBox anInterface = new JCheckBox("Interface");
            add(anInterface);
            add(new JButton(LangUtil.i18n("button.search")) {{
                addActionListener(e -> {


                    if (StringUtil.isBlank(owner.getText()) && StringUtil.isBlank(name.getText()) && StringUtil.isBlank(description.getText()) && !anInterface.isSelected()) {
                        MessageUtil.warning("Please input");
                        return;
                    }



                    JavaOctetEditor.getInstance().task
                            .submit(new SearchMethodTask(JavaOctetEditor.getInstance().jar, owner.getText(), name.getText(), description.getText(), anInterface.isSelected()))
                            .thenAccept(it -> {
                                ((DefaultListModel<ResultNode>) resultList.getModel()).clear();
                                for (ResultNode resultNode : it) {
                                    ((DefaultListModel<ResultNode>) resultList.getModel()).addElement(resultNode);
                                }
                            });
                });
            }});
        }}, BorderLayout.SOUTH);
    }
}

File: ui/src/main/kotlin/cn/enaium/joe/ui/dialog/SearchMethodDialog.kt
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.ui.dialog

import cn.enaium.joe.core.task.SearchMethodTask
import cn.enaium.joe.ui.Instance.jar
import cn.enaium.joe.ui.JavaOctetEditor.Companion.task
import cn.enaium.joe.ui.control.SearchResultList
import cn.enaium.joe.ui.util.i18n
import cn.enaium.joe.ui.util.warning
import javafx.scene.control.Button
import javafx.scene.control.CheckBox
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.scene.layout.BorderPane
import org.tbee.javafx.scene.layout.MigPane

/**
 * @author Enaium
 * @since 2.0.0
 */
class SearchMethodDialog : Dialog(i18n("search.method.title")) {
    init {
        content = BorderPane().apply {
            val searchResultList = SearchResultList()
            top = MigPane("fillx", "[fill][fill]").apply {
                add(Label(i18n("search.owner")))
                val owner = TextField()
                add(owner, "wrap")
                add(Label(i18n("search.name")))
                val name = TextField()
                add(name, "wrap")
                add(Label(i18n("search.description")))
                val description = TextField()
                add(description, "wrap")
                add(Label(i18n("search.interface")))
                val anInterface = CheckBox()
                add(anInterface, "wrap")
                add(Button(i18n("menu.search")).apply {
                    setOnAction {
                        if (owner.text.isBlank() && name.text.isBlank() && description.text.isBlank() && !anInterface.isSelected) {
                            warning("Please input!")
                            return@setOnAction
                        }
                        searchResultList.items.clear()
                        task.submit(
                            SearchMethodTask(
                                jar,
                                owner.text,
                                name.text,
                                description.text,
                                anInterface.isSelected
                            )
                        )
                            .thenAccept {
                                it.forEach { result ->
                                    searchResultList.items.add(result)
                                }
                            }
                    }
                }, "span 2")
            }
            center = searchResultList
        }
    }
}